src/buffer.rs

```diff
@@ -357,3 +368,66 @@ pub enum EncMacroblockParameterBuffer {
     /// Abstraction over `VAEncMacroblockParameterBufferH264`
     H264(h264::EncMacroblockParameterBufferH264),
 }
+
+/// Helper to access a single segment of mapped coded buffer
+pub struct CodedBufferSegment<'s> {
+    pub bit_offset: u32,
+    pub status: u32,
+    pub buf: &'s [u8],
+}
+
+/// Helper to access segments of mapped coded buffer
+pub struct CodedBuffer<'p> {
+    segments: Vec<CodedBufferSegment<'p>>,
+    buffer: &'p Buffer,
+}
+
+impl<'p> CodedBuffer<'p> {
+    /// Map a 'VAEncCodedBufferType' buffer. `picture` makes sure that the surface was synced and
+    /// would not change state until this buffer is unmapped (dropped).
+    pub fn new(buffer: &'p Buffer) -> Result<Self, VaError> {
+        let mut addr = std::ptr::null_mut();
+        let mut segments = Vec::new();
+
+        va_check(unsafe {
+            bindings::vaMapBuffer(buffer.context.display().handle(), buffer.id, &mut addr)
+        })?;
+
+        while !addr.is_null() {
+            let segment: &bindings::VACodedBufferSegment =
+                unsafe { &*(addr as *const bindings::VACodedBufferSegment) };
+
+            let size = segment.size;
+            let buf = segment.buf;
+
+            let buf = unsafe { std::slice::from_raw_parts(buf as *mut u8, size as usize) };
+
+            segments.push(CodedBufferSegment {
+                bit_offset: segment.bit_offset,
+                status: segment.status,
+                buf,
+            });
+
+            addr = segment.next;
+        }
+
+        Ok(Self { segments, buffer })
+    }
+
+    /// Returns the segments of mapped coded buffers.
+    pub fn segments(&self) -> &Vec<CodedBufferSegment<'p>> {
+        &self.segments
```

Comment:
How about implementing `Iterator` instead?

Reply:
I added the `iter` function. Or should I add a cursor logic to `MappedCodedBuffer`? :smile:

Reply:
`iter` is good! I suspect we can remove `segments`, but let's do that in a follow-up CL if needed.